### PR TITLE
35 fix mavenbridgeimpl

### PR DESCRIPTION
**Pull request by jan-vcapgemini**
First test using dependencyManagement and springboot framework.
Could be simplified in next iteration.

### CODE REVIEW

File: java/impl/src/test/java/net/sf/mmm/code/impl/java/JavaExtendedContextWithMavenAndSourceCodeTest.java [hunk 1]
```diff
@@ -86,7 +86,8 @@ public void testContextFromOwnMavenProject() {
 
     // then
     assertThat(type.getContext()).isSameAs(context);
-    assertThat(type.getDoc().getLines()).contains("Abstract top-level interface for any item of code as defined by this API. It reflects code structure.");
+    assertThat(type.getDoc().getLines()).contains(
+        "Abstract top-level interface for any item of code as defined by this API. It reflects code structure.");
```
Comment (hohwille) [inline java/impl/src/test/java/net/sf/mmm/code/impl/java/JavaExtendedContextWithMavenAndSourceCodeTest.java:90]: is there any real change in this file? I only see that you are using a different line-length in the formatter and therefore all changes are just formatter chnages in this file...

File: java/maven/src/test/java/net/sf/mmm/code/java/maven/impl/MavenBridgeImplTest.java [hunk 2]
```diff
@@ -24,6 +24,11 @@
 
   private static final File POM_XML = new File(MavenConstants.POM_XML);
 
+  /**
+   * Root Path where to test data is stored
+   */
+  private static final File rootTestPath = new File("src/test/resources/testdata/");
```
Comment (hohwille) [inline java/maven/src/test/java/net/sf/mmm/code/java/maven/impl/MavenBridgeImplTest.java:30]: please follow Java coding conventions and name `ROOT_TEST_PATH` instead.

File: java/maven/src/test/java/net/sf/mmm/code/java/maven/impl/MavenBridgeImplTest.java [hunk 3]
```diff
@@ -24,6 +24,11 @@
 
   private static final File POM_XML = new File(MavenConstants.POM_XML);
 
+  /**
+   * Root Path where to test data is stored
+   */
```
Comment (hohwille) [inline java/maven/src/test/java/net/sf/mmm/code/java/maven/impl/MavenBridgeImplTest.java:29]: gramatically incorrect and IMHO not explaining anything more what is not already obvious by its value.
IMHO just delete this JavaDoc.

File: java/maven/src/test/java/net/sf/mmm/code/java/maven/impl/MavenBridgeImplTest.java [hunk 4]
```diff
@@ -48,6 +53,24 @@ public void testReadModel() {
...
+
+    // given
+    File mavenProjectDirectory = new File(rootTestPath, "localmavenproject/maven.project/core"); // test Maven project
+
+    // when
+    MavenBridgeImpl reader = new MavenBridgeImpl();
+    Model model = reader.readEffectiveModel(new File(mavenProjectDirectory, "pom.xml"));
+
+    // then
+    assertThat(model).isNotNull();
```
Comment (hohwille) [inline java/maven/src/test/java/net/sf/mmm/code/java/maven/impl/MavenBridgeImplTest.java:71]: shouldnt this test assert something more reasonable in addition?
However, surely agreed that the POM shall be parsable and therefore the model shall not be null. If that failked before sufficient for the PR to be acceptable :)

**Review by hohwille — COMMENTED** [review]
I am not quite sure what is the intention of this PR.
It adds some maven project and a JUnit but no chode changes to MavenBridgeImpl.
However the PR title is "35 fix mavenbridgeimpl".
This is slightly confusing...

OK, seems this is WIP. Maybe you want to check if your test reveals the bug via JUnit. I approved the github actions to run on this PR.

**Comment by hohwille** [comment]
mhm, so all checks passed. JUnit does not reveal any bug and no fix is included.

**Comment by jan-vcapgemini** [comment]
I've found out that the issue only occurs, when a .mvn folder with a maven.config file and following content is present:
`-Drevision=1.0.0-SNAPSHOT`

The resolveProperties method completely overwrites the System Properties with the revision property.
https://github.com/m-m-m/code/blob/0470b864e3077d3d439dcfed87b8be8fdcd76a0e/java/maven/src/main/java/net/sf/mmm/code/java/maven/impl/MavenBridgeImpl.java#L217-L221

**Comment by jan-vcapgemini** [comment]
I've added a maven.config to the test project and merged all of @maybeec 's changes to this PR.
The test succeeds now and the issue seems to be fixed.

**Comment by maybeec** [comment]
Please charge it to be non draft PR now

**Review by hohwille — APPROVED** [review]
The finding and loading of `maven.config`  now looks sane. 
Still there are many pointless format changes, however due to importance of the bug, I am happy to merge as is. Thanks @jan-vcapgemini  and @maybeec for this great contribution 👍

**Review by hohwille — CHANGES_REQUESTED** [review]
Tests failing:
```
Error:  Failures: 
Error:    MavenBridgeImplTest.testReadModel:51 
expected: "${revision}"
but was : null
Error:    MavenBridgeImplTest.testReadProject:93->verifyDependencies:131 Expected
 [Dependency {groupId=org.apache.maven, artifactId=maven-core, version=3.6.1, type=jar}, Dependency {groupId=net.sf.m-m-m, artifactId=mmm-util-io, version=7.5.1, type=jar}, Dependency {groupId=net.sf.m-m-m, artifactId=mmm-util-test, version=7.5.1, type=jar}, Dependency {groupId=org.junit.jupiter, artifactId=junit-jupiter, version=5.7.0, type=jar}, Dependency {groupId=org.assertj, artifactId=assertj-core, version=3.19.0, type=jar}, Dependency {groupId=ch.qos.logback, artifactId=logback-classic, version=1.3.0-alpha5, type=jar}]
to contain
  [Dependency {groupId=net.sf.m-m-m, artifactId=mmm-util-io, version=7.5.1, type=jar}, Dependency {groupId=net.sf.m-m-m, artifactId=mmm-util-test, version=7.5.1, type=jar}, Dependency {groupId=org.apache.maven, artifactId=maven-core, version=3.6.1, type=jar}]
but did not expect
  [Dependency {groupId=ch.qos.logback, artifactId=logback-classic, version=1.3.0-alpha5, type=jar}, Dependency {groupId=org.junit.jupiter, artifactId=junit-jupiter, version=5.7.0, type=jar}, Dependency {groupId=org.assertj, artifactId=assertj-core, version=3.19.0, type=jar}]
Error:  Errors: 
Error:    MavenBridgeImplTest.testResolveDependencyManagementVersionFromParent:68 » IllegalState
[INFO] 
Error:  Tests run: 4, Failures: 2, Errors: 1, Skipped: 0
```

**Comment by hohwille** [comment]
There are two test errors:
1. https://github.com/m-m-m/code/blob/c9bde3af9ba82bf5dd992bca51043f89bdd16b15/java/maven/src/test/java/net/sf/mmm/code/java/maven/impl/MavenBridgeImplTest.java#L51
2. https://github.com/m-m-m/code/blob/c9bde3af9ba82bf5dd992bca51043f89bdd16b15/java/maven/src/test/java/net/sf/mmm/code/java/maven/impl/MavenBridgeImplTest.java#L93 which is verifying the dependencies of the module itself which is https://github.com/m-m-m/code/blob/c9bde3af9ba82bf5dd992bca51043f89bdd16b15/java/maven/pom.xml

**Comment by maybeec** [comment]
are pr builds blocked by intention? It would have been valuable to see directly the test case failures on GA at push time to not delay further the work on this topic.

**Comment by hohwille** [comment]
> are pr builds blocked by intention?

Nope. Seems to be a newer github feature. I did not enable this and could not find any option in the settings to change it either.
There is `Fork pull request workflows from outside collaborators` in `Settings  > Actions` but it gives me three options and none of them allows to prevent approval. Maybe cause there is no billing plan associated with my org...

**Comment by maybeec** [comment]
https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks

**Comment by hohwille** [comment]
I just had a couple of minutes but 1. is not caused by this PR for sure.
Version is only present in the parent
https://github.com/m-m-m/code/blob/c9bde3af9ba82bf5dd992bca51043f89bdd16b15/java/maven/pom.xml#L8
And this seems to read the raw model without inheritance and interpolation.
Maybe this also changed with the maven update and dependency changes lately.
Version was earlier present in pom itself but never just as `${revision}`:
https://github.com/m-m-m/code/blob/eaf3980db2baa147acd77e48c33a83763944e6d3/java/maven/pom.xml#L12

So why are the tests not failing on master?
```
[INFO] Tests run: 0, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- maven-jar-plugin:3.1.2:jar (default-jar) @ mmm-code-java-maven ---
[INFO] Building jar: /home/runner/work/code/code/java/maven/target/mmm-code-java-maven-1.0.0-beta7-SNAPSHOT.jar
```

Seems that no test is executed there. Also currently not yet any clue why... Need some time but have none. Pity.

**Comment by maybeec** [comment]
Test did not run as @jan-vcapgemini  already told above, it seemed to be an issue of the migration to junit5 in your code. It was fixed by him in this PR, therefore you see the issue raising here right now @hohwille. I think @jan-vcapgemini  can shortly have a look into these issues and clarify them to help out.

**Comment by hohwille** [comment]
> https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks

That is exactly what I found but it does not help in any way. If you find a concrete option that would prevent this manual approval I would be happy but there is none.

**Comment by hohwille** [comment]
fixing 1. is easy:
https://github.com/m-m-m/code/blob/c9bde3af9ba82bf5dd992bca51043f89bdd16b15/java/maven/src/test/java/net/sf/mmm/code/java/maven/impl/MavenBridgeImplTest.java#L51
Change to:
```
assertThat(model.getParent().getVersion()).isEqualTo("${revision}"); 
```
And we are done.
IMHO 2. is more tricky as I would first like to understand what triggered this change. Seems that when this test was written no transitive dependencies have been included and this now changed. If that is confirmed, we could change `verifyDependencies` to accept a flag that tells that it should check for `contains` rather than `containsExactly` and set this flag to `true` for `testReadProject` but leave it as `false` for `testReadModel`.

**Comment by hohwille** [comment]
For the GHA workflows, I will follow devonfw best practices and separate CI from PR builds.

**Comment by jan-vcapgemini** [comment]
The testReadProject() verifyDependencies method is a bit too restrictive I think.
https://github.com/m-m-m/code/blob/8bf499fe980220ff1b33d57d2029eb0c8928bbcc/java/maven/src/test/java/net/sf/mmm/code/java/maven/impl/MavenBridgeImplTest.java#L70-L72

I've adjusted it now to fix this PR but it should be checked again as it might fail with other changes.

**Comment by jan-vcapgemini** [comment]
There is still an issue with the resolveProperties method. Currently the revision of the test project gets overwritten with the version number of the main project.

**Comment by hohwille** [comment]
> 
> 
> There is still an issue with the resolveProperties method. Currently the revision of the test project gets overwritten with the version of the main project.

This is not a bug of mmm-code but the same behaviour that maven has on commandline:

```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for maven.project 1.0.0-beta7-SNAPSHOT:
[INFO]
[INFO] maven.project ...................................... SUCCESS [  3.663 s]
[INFO] core ............................................... SUCCESS [  6.665 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  11.268 s
[INFO] Finished at: 2021-11-18T08:50:46+01:00
[INFO] ------------------------------------------------------------------------
```

**Comment by hohwille** [comment]
However, what confuses me is the "fix" for reading `.mvn/maven.config` by @maybeec.

https://github.com/m-m-m/code/blob/a990208101354eb6838328183f244867a55ac259/java/maven/src/test/resources/testdata/localmavenproject/maven.project/pom.xml

There is no `pom.xml` file in its parent folder. So I thought it should not read the `.mvn/maven.config`.
However, it only prevents reading it when there is a `pom.xml` in the parent folder, but still it otherwise traverses up to the root folder of the filesystem.
I thought the only thing for a correct "fix" would have been to stop traversing after the first `.mvn/maven.config` has been found and read. Aint this the behaviour of the official maven?

**Comment by hohwille** [comment]
Mhm, maven's behaviour still seems to be different to what I thought:
https://github.com/apache/maven/blob/2c37a7a38e8a4bb2b60e67d3169d4e660f477abf/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java#L393

**Comment by maybeec** [comment]
> Mhm, maven's behaviour still seems to be different to what I thought:
https://github.com/apache/maven/blob/2c37a7a38e8a4bb2b60e67d3169d4e660f477abf/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java#L393

That's actually a new behavior of maven 3.6 I think to get to know the root of a multi module project and it's based on the location of the .mvn folder: https://github.com/apache/maven/blob/46410d3df2458af9b2118ace6cc1a68b75c5d262/.mvn/readme.txt
So you have a loop in your arguments

**Review by hohwille — APPROVED** [review]
@jan-vcapgemini thanks for all your updates, cleanup and polishing. Looks perfect to me now. 👍 Going to merge and plan to push out a new release at least on weekend.

**Comment by hohwille** [comment]
unable to build the release for now:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-gpg-plugin:1.6:sign (sign-artifacts) on project mmm-code-parent: Exit code: 2 -> [Help 1]
```

**Pull request merged**